scalding-core/src/main/scala/com/twitter/scalding/Execution.scala

```diff
@@ -192,6 +193,8 @@ sealed trait Execution[+T] extends java.io.Serializable {
    */
   def zip[U](that: Execution[U]): Execution[(T, U)] =
     Zipped(this, that)
+
+  override val hashCode = super.hashCode
```

Comment:
Okay, reworked this with `ScalaRunTime`, also to preserve the `hashCode`/`equals` contract.  Note that I've introduced a few more bits to expand the cases where we can reconstruct and achieve equality.

Also added a bunch of tests around reconstruction and the handling of mutable `Execution`s.